core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,653 @@
+   * cancellation and is it's own independent root for cancellation.
+   */
+  public CancellableContext fork() {
+    return new Context(this).withCancellation();
+  }
+
+  /**
+   * Attach this context to the thread and make it {@link #current}, the previously current context
+   * will be restored when detach is called. It is allowed to attach contexts where
+   * {@link #isCancelled()} is {@code true}.
+   */
+  public void attach() {
+    contextStack.get().addLast(this);
+  }
+
+  /**
+   * Returns {@code true} if this context is {@link #current}, {@code false} otherwise.
+   */
+  public boolean isCurrent() {
+    return current() == this;
+  }
+
+  /**
+   * Detach the current context from the thread and restore the context that was previously
+   * attached to the thread as the 'current' context.
+   *
+   * @throws java.lang.IllegalStateException if this context is not {@link #current()}.
+   */
+  public void detach() {
+    ArrayDeque<Context> stack = contextStack.get();
+    if (stack.isEmpty()) {
+      if (this == ROOT) {
+        throw new IllegalStateException("Cannot detach root");
+      } else {
+        throw new IllegalStateException("Cannot detach non-root context when root is current");
+      }
+    }
+    if (stack.peekLast() != this) {
+      throw new IllegalStateException("Cannot detach a context that is not current");
+    }
+    stack.removeLast();
+  }
+
+  /**
+   * Is this context cancelled.
+   */
+  public boolean isCancelled() {
+    if (parent == null || !cascadesCancellation) {
+      return false;
+    } else {
+      return parent.isCancelled();
+    }
+  }
+
+  /**
+   * Return the cause of the cancellation or {@code null} if context was cancelled without a cause.
+   */
+  @Nullable
+  public Throwable cause() {
+    if (parent == null || !cascadesCancellation) {
+      return null;
+    } else {
+      return parent.cause();
+    }
+  }
+
+  /**
+   * Add a listener that will be notified when the context becomes cancelled.
+   */
+  public void addListener(final CancellationListener cancellationListener,
+                          final Executor executor) {
+    Preconditions.checkNotNull(cancellationListener);
+    Preconditions.checkNotNull(executor);
+    if (canBeCancelled()) {
+      ExecutableListener executableListener =
+          new ExecutableListener(executor, cancellationListener);
+      synchronized (this) {
+        if (isCancelled()) {
+          executableListener.deliver();
+        } else {
+          if (listeners == null) {
+            // Now that we have a listener we need to listen to our parent so
+            // we can cascade listener notification.
+            listeners = new ArrayList<ExecutableListener>();
+            if (parent.canBeCancelled()) {
+              parent.addListener(parentListener, MoreExecutors.directExecutor());
+            }
+          }
+          listeners.add(executableListener);
```

Comment:
This add needs to be done before `parent.addListener()`. There is a race between `isCancelled()` and `parent.addListener()` such that `addListener()` notifies `parentListener` immediately.